_posts/2020-01-21-flyway-spring-boot-multitenancy.md

```diff
@@ -0,0 +1,242 @@
+        return DataSourceBuilder.create()
+                .url(source.get("jdbcUrl"))
+                .driverClassName(source.get("driverClassName"))
+                .username(source.get("username"))
+                .password(source.get("password"))
+                .build();
+    }
+}
+```
+Now we can define the data source configuration for multiple `DataSource`s.
+
+```java
+@Configuration
+public class DataSourceConfiguration {
+
+    private final DataSourceProperties dataSourceProperties;
+
+    public DataSourceConfiguration(DataSourceProperties dataSourceProperties) {
+        this.dataSourceProperties = dataSourceProperties;
+    }
+
+    @Bean
+    @PostConstruct
+    public DataSource dataSource() {
+        CarRoutingDataSource customDataSource = new CarRoutingDataSource();
+        customDataSource.setTargetDataSources(dataSourceProperties.getDatasources());
+        return customDataSource;
+    }
+}
+```
+
+We have only one `DataSource` in this configuration and it should be always the `DataSource` of the tenant from the
+request. We can achieve this by using `AbstractRoutingDataSource`. `AbstractRoutingDataSource`
+can manage multiple `DataSource`s and routes over connections. We can extend `AbstractRoutingDataSource`
+to route over our Car `Datasource`s.
+
+```java
+public class CarRoutingDataSource extends AbstractRoutingDataSource {
+
+    private static final String DEFAULT_TENANT = "vw";
+
+    @Override
+    protected Object determineCurrentLookupKey() {
+        return ObjectUtils.defaultIfNull(ThreadTenantStorage.getTenantName(), DEFAULT_TENANT);
+    }
+}
+````
+The `CarRoutingDataSource` will call `determineCurrentLookupKey` whenever it gets a new connection.
+The current tenant has been set to `ThreadTenantStorage`, so the method `determineCurrentLookupKey `
+returns this current tenant, the `CarRoutingDataSource` find the `DataSource` of this tenant and set it as current
+`DataSource` automatically. It means every access to data from the thread will be routed to this `Datasource`.
+
+## Migration of SQL scripts with multiple data sources
+
+If we want to have version control over the database with Flyway and make changes in it like adding a column, adding a table,
+dropping a constraint and so on, we have to write SQL script. With the Flyway support for Spring Boot we just need
+to deploy the application and new scripts are migrated. By default, Flyway uses the current `DataSource`.
+But we want to apply the scripts to all `DataSource`s we defined in the application.
+We didn't define static `Datasource`s, but we put them in a `Map` in the `DataSourceProperties`, so now we can iterate
+over them.
+
+```java
+@Configuration
+public class DataSourceConfiguration {
+
+    private final DataSourceProperties dataSourceProperties;
+
+    public DataSourceConfiguration(DataSourceProperties dataSourceProperties) {
+        this.dataSourceProperties = dataSourceProperties;
+    }
+
+    @Bean
+    @PostConstruct
```

Comment:
Ich verstehe nicht, wie das `@PostConstruct` und `@Bean` zusammen funktionieren. Das `PostConstruct` soll vermutlich die Flyway-Migration ausfuehren und das `Bean` soll die `DataSource` bereitstellen. 

Aaaah...in deinem echten Code sieht das auch anders aus. Ich wuerde trotzdem das `@Bean` und das `@PostConstruct` komplett trennen und jeweils in eine eigene Methode packen. Das macht es sauberer und leichter zu verstehen.

Reply:
Guter Punkt. Ich habe die Konfiguration von `DataSources` und die Flyway-Migration getrennt